### PR TITLE
Circleci suggestions

### PR DESCRIPTION
**Pull request by zmarkan**
Work in progress:
- replaced codecov job with the commands in the orb
- aliases with executors (docker, machine) and gradle cache restoration command
- parameterised ui tests and removed the dublicate job definition for oss and gms

TODO: 
- Save build cache

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -2,38 +2,43 @@
...
-  - &restore_gradle_cache
-    restore_cache:
-      name: Restore gradle cache
-      keys:
-        - gradle-v3-{{ arch }}-{{ checksum "/tmp/gradle_cache_seed" }}
-        - gradle-v3-{{ arch }}-
-        - gradle-v3-
-  - &android-docker-image
-    image: cimg/android:2022.04
+commands:
```
Comment (growse) [inline .circleci/config.yml:5]: This is going under "A thing I learned today". Nice.

**Comment by zmarkan** [comment]
@growse entirely up to you, happy to do either way. For now I'm planning to do a few more tweaks to the flow etc & have it ready for you to review later today

**Comment by growse** [comment]
LGTM, only request is that these get rebased and squashed before merge.

On the gradle caching point, I think the lack of clarity was around the fact that only some of the original jobs were saving a gradle cache. This was somewhat deliberate - you can't overwrite a cache with the same key. This meant that in a workflow where the `build` job ran followed by the `lint` job, there's no point having the lint job save the cache, because it'll try and overwrite the cache saved by the build job. This is one of the reasons there's so many building gradle tasks actually being executed in the build job, in order to maximize the efficiency of the cache for later jobs (that run lint, unit tests etc). So I'd suggest the `save-gradle-cache` steps on lint & unit-test are probably redundant.

Separately, I've added in a smoketest annotation on master in 2f28e00f, so these can be invoked on the PR job with:

```
./project/gradlew -p project/ connectedGmsDebugAndroidTest -Pandroid.testInstrumentationRunnerArguments.annotation=org.owntracks.android.testutils.SmokeTest
```

**Comment by zmarkan** [comment]
Totally agree re squashing & rebasing - that's why I'm keeping it as a draft PR for now while I'm hacking on it - 
Fair point re caching - 

And as for smoke tests - I was just doing the exact same thing separately so I'll rebase onto your branch & use that - thanks for letting me know!

**Comment by zmarkan** [comment]
Ok so I think this is now done. Marking this as ready for review.
I've tested the dynamic workflow on my own fork and it works fine, just runs the smoke tests in most cases, before running the rest.

One other idea I had about caching is to provide multiple cache keys in varying specificity - basically split it into dependencies/build script, and code caches, something along the lines of:

```yaml
    steps:
      - restore_cache:
          keys:
            # Find a cache corresponding to this specific package-lock.json checksum
            # when this file is changed, this key will fail
            - $PREFIX-gradle-deps-$DEPS_CHECKSUM-build-cache-BUILD_CACHE_CHECKSUM
            - $PREFIX-gradle-deps-$DEPS_CHECKSUM-*
            - $PREFIX-*
```

That way you get _some_ benefit from the cache at least for the deps, even if the sources have changed. I've not tried it yet but I can give it a go in a separate PR.

**Comment by growse** [comment]
I was using a hierarchy of cache keys on the current version (https://github.com/owntracks/android/blob/master/.circleci/config.yml#L12). I think I was originally using the `restore-gradle-cache` command in the `android` orb, but then broke it apart to support the hierarchy.

If you could adjust the restore-cache part to do that, I think this is good to merge.

**Comment by growse** [comment]
Nice idea on using different cache seed segments based on different file lookups.

If you can squash to a single, I can rebase / merge. Thanks

**Comment by zmarkan** [comment]
Cool, done.

**Pull request merged**